### ISSUE TITLE
D(1) no longer produces a trace

### ISSUE THREAD

**Issue by raveren** [kint-php/kint#368]
<img width="536" alt="image" src="https://user-images.githubusercontent.com/364220/133849939-7c0937de-f7ae-447e-b9ef-0be3607d57ad.png">

<img width="252" alt="image" src="https://user-images.githubusercontent.com/364220/133849807-0706def7-3eef-4cf6-a12e-c078dd929be4.png">

please don't tell me you removed it?

**Comment by jnvsor** [comment]
I did, in 6c4505e238fc13b423ea7657c23e68437aa8b914

It was a messy bit of code and only worked if the call was found correctly. I had a lot of annoying situations where I didn't know where the code was being called and couldn't find out with `d(1)` because it was stuck in an `eval` or a `call_user_func` or something and kint couldn't detect the call to see if it was a literal `1` or a variable, so instead of a stack trace it'd just say "integer 1" and I'd have to explicitly call trace and try again

Worst of all: It generated bogus access paths because it removed php internal stack frames before dumping, shifting indices for part of the stack

Since the changes in 01d6ba37ec24d020730fdf6daeaea03c91e35f98 it's now possible to override internal `Kint` class methods because all the calls are `static` instead of `self` with the min PHP version bump to 5.6. In other words, you can override to make `d(1)` work again yourself, or just add a `t()` helper would probably be faster

I see I forgot to update the readme...

**Comment by raveren** [comment]
so you removed a wonderful piece of functionality because you didn't like that it produced messy code :))

no pull requests from the community for years and no merged PRs (since when you rewrote everything) could _maybe_ indicate the code is not as readable as you would think in the first place.

please add the helper or do something, i have to type `\Kint::trace();dd();` each time now that I want to see trace, it's horrible

**Comment by jnvsor** [comment]
> or do something

The question is what

I was thinking of maybe using `Kint::__call()` and letting the user publish aliases to the global namespace manually and in the end it was an over-engineered pile of rubbish...

Then I thought we could generate global functions in code but I can't find any way to do that without heavy `eval` use so that's a no go...

We might be able to make a sort of "Build your own" kint download where you can click together which helpers you want and what to name them but again: over-engineering...

And I still don't want to pollute the global namespace. Half the l10n stuff out there probably uses `t()` for translation, so that's _bound_ to cause complaints

---

So this is the best idea I've come up with so far: We add a separate package `kint-php/kint-helpers` and load that full of any extra helpers we want.

* We can put anything in there without cluttering up kint
  * And this helps kints tests, which are exponentially more complex the more helpers we have
  * So `dd()`, `ddd()`, `s()`, `sd()`, `t()`, `td()` for starters? Or should we make the trace alias `tr()`?
* We don't have to worry about the global namespace, because the user has to explicitly install it
  * We might have to add a way to disable them one by one, probably with the `Utils::composerGetExtras()` thing
* Since it'll require kint you can just require the helpers instead and it'll give you the whole shebang
* If someone only wants one of them they can just copy-paste the function straight from the helper package to their own code (And we can link to the repo from where the site demos helpers now)

Then we just add it to suggests on the main package and we're done. What do you think about this idea @raveren ?

We still need a place to draw the line, otherwise I can see `prnalfd()` enabling "plaintext return no-depth-limit aante-light-theme folder die dumps" becoming one of hundreds of aliases... A function for every use case! We need to draw the line somewhere, and I don't see a clear place to do that

> no pull requests from the community for years and no merged PRs (since when you rewrote everything) could maybe indicate the code is not as readable as you would think in the first place

Could be. I like to think it's because it's a lot easier to change functionality without having to touch `vendor` now. People probably just tweak things themselves. If you see any parts you think could use a clean up let me know

Packagists stats show a good 10k installs a day so it's not for lack of use anyway

**Comment by raveren** [comment]
Hey! Good to hear from you, I'm glad to see you're still enthusiastic about Kint, so am I - but I have had chronically little time to devote to hobbies this past year since I'm back to programming.

It's a nice idea to separate the potentially incompatible global helpers to a new repo. However I feel like it's the typical old case of bad desktop software where the programmer provides many options for the user, when instead they should have just built a great experience out of the box. The configuration options may be there for the advanced users, but the 99.999% just want a `dd` with some definite benefits over symfonys varDumper. Of which Kint undoubtedly has several, but still it's been invisible to PHP-centric media for years - and it's unfair. Kint is great and the world deserves to use it :)

What Kint is definitely **lacking** since I've been using it a year now again:

* dd(1) - this was an amazing "wow" feature
* https://gist.github.com/raveren/eba373d8abb572b0528c73d145103f95
  * kd() (instead of dd()) - MUST have, THE reason no one cares about kint anymore - it does not have dd anymore.
  *  there's an `ss` which should have been `s` - when I'm working in the console I love the feature to display some dumps without colors to differentiate them instantly for the regular ones. As it is now - s and d are identical in the console.
  * edit: I also love having `$debugInfo = @ss($arg);` when testing as message on assertion fail.
  * this config still uses the Remotecall plugin to link to PHPStorm. However there's already (10+ years later) a phpstorm:// protocol that **almost** works for this. Remotecall is still better, but the protocol comes out of the box, so Kint should be pre-configured for it.
  * `Kint\Renderer\RichRenderer::$folder = false;` no idea whether this is still in the HEAD or not, but that was horrible.

And if you'd like Kint to get picked up by the media, become successful and viral, there should be a laravel integration plugin, if you could point me into the right direction on how to achieve such a thing, I could cook it up in a couple of months, it should:
 * Filter out classes residing in `/vendor/` from the trace
 * Call ->toArray() for eloquent instances

And it's already a useable experience for working with laravel, which is all the rage nowadays in the PHP world. I use it too (and have to admit it's pretty great)

**Comment by jnvsor** [comment]
> `dd(1)`

Are we talking about the `dd` part or the `1` part?

One of the reasons I'm particular about the global ns is because this already existed in laravel, and then depending on the order of requires and/or other stuff one of the two would be unable to be declared. That's why you had to make `ddd()`. Maybe we should call it "dump and exit" instead of "dump and die" and then make it `de()`?

If it's the trace you're after we could make a `trd()` helper that does the same as the old `dd(1)`

Still not convinced we shouldn't just throw the helpers into their own package though

> when I'm working in the console I love the feature to display some dumps without colors to differentiate them instantly for the regular ones. As it is now - s and d are identical in the console.

Well you can use the `~` modifi-

> I also love having `$debugInfo = @ss($arg);` when testing as message on assertion fail.

Ah. Yeah that makes sense. I kind of want to say do `~~@d($arg)` but that's hardly intuitive hehe. Though I do think a separate helper just to return in plaintext is an example of a helper that's too over-specific

> this config still uses the Remotecall plugin to link to PHPStorm

`$file_link_format` right? I don't use it myself so I wouldn't know what is the best way to do it. (I tried getting it working with subl once but gave up) I've kept the `ini_get('xdebug.file_link_format')` default setting you originally had.

I'd like to keep it generic if possible, but if you know any new standards or something that is more widely used I'll switch to that as the default

> no idea whether this is still in the HEAD or not, but that was horrible

Yeah I agree, that was a mistake on my part (And it's another reason why I don't want to add overly opinionated stuff)

> And if you'd like Kint to get picked up by the media, become successful and viral, there should be a laravel integration plugin

I don't really care if it goes viral. I just want it to be good :)

> I could cook it up in a couple of months

Should only take a few hours but I say that as someone who knows the code... Interested to see how long it takes you in practice

> Filter out classes residing in /vendor/ from the trace

Depends how you want to handle it, but assuming you don't want to screw with traces that are dumped explicitly I'd subclass and replace `Kint\Parser\TracePlugin` and only remove those frames if the array is depth 0 and named `Kint\Kint::trace()`

> Call ->toArray() for eloquent instances

This is basically the exact scenario the [writing plugins](https://kint-php.github.io/kint/writing-plugins/) demo shows. You can add an `if($var instanceof ...) return;` at the top and be halfway there.

Unfortunately that page still references v3 code and I don't want to update the docs until 4.x goes into beta at least. I've been wrestling with gh actions over php8.1, but that's my last hurdle for beta

I don't want to add new features after it hits beta, so if we're deciding what to do about the helpers we have a 2 month deadline.

I still think we should make a separate helper package. Then we can go ham with `ss()` and anything else we feel like and we sidestep the inevitable bike-shedding over which ones are important enough to add by default (And I can't stress enough how much of a pain it is to integration test kint helpers, not to mention the ones that die at the end would kill the test suite!)

Here's a draft of potential extra helpers:

* dd
* ddd
* tr
* trd
* s
* sd
* ss
* ssd

Anything else you can think of?

**Comment by raveren** [comment]
Listen, lets be real for a moment. This is just another (counting at least fourth or fifth) clear and outright regression in functionality since I left Kint into your hands, and it's getting ridiculous. 

Please revert the removal of dd(1) or I will do it.

And as for the laravel plugin:

> Depends how you want to handle it

I see there's no possibility for me to create it, I tried to read your code, but, sorry, but it's a confusing mess, there's no way I can change anything there without spending days to comprehend what is happening there.

**Comment by jnvsor** [comment]
Look. I'll take feedback, and I'll take criticism, and I'll change things when it's clear I made a mistake (eg. folder) but I'm not going to take ultimatums or threats from anyone.

**Comment by kktsvetkov** [comment]
> We add a separate package `kint-php/kint-helpers` and load that full of any extra helpers we want.

That's a good suggestion, it's the best of both worlds -- both the main project is clean of the global functions, and yet people that are used to the helper functions can just use that.

**Comment by kktsvetkov** [comment]
> Maybe we should call it "dump and exit" instead of "dump and die" and then make it `de()`?

How about `x()`? I don't think there are any clashes with such a global function, and `x` is for _e**X**it_.

**Comment by jnvsor** [comment]
@kktsvetkov Hi, I've added a repo with a preliminary version of it. Would you like to try it out? It's a single file s you can just require that and call it a day: https://github.com/kint-php/kint-helpers/blob/master/init.php

If there are no further complaints or suggestions I'll add it to packagist

**Comment by DRSDavidSoft** [comment]
@jnvsor Late to the party, but in all these times I've been wondering why not use one of the modifiers in  order to exit the execution? e.g. `!d(....)` wouldn't it make more sense, as it could replace other shorthand  + exit functions as well? e.g. `!s(....)` and etc

**Comment by DRSDavidSoft** [comment]
Also, I was wondering what was the difference between `dd()` and `ddd()`, curious to know.  When `dd()` was deprecated, I simply did a regex replace in my code-base from `dd()` to `ddd()`, but never checked the difference between the two.

**Comment by jnvsor** [comment]
@DRSDavidSoft They were the same. It was mostly because at the time laravel was the only other one that had a `dd()` and the conflict was making early versions of either it or kint blow up. This was all a loooong time ago though, and now that symfony/var-dumper has `dd()` as well there's not much point making a `ddd()` when it's a de facto industry standard

> why not use one of the modifiers in order to exit the execution

Well firstly because that'd be a breaking change, secondly because modifiers aren't perfect. In the event kint can't determine which call it was modifiers will be ignored.

For instance, if you have 2 "identical" calls on the same line:

```php
d($a); !d($b); // Kint can't tell which call it's in so opts for the safe route of assuming nothing
```

**Comment by DRSDavidSoft** [comment]
Understandable, I wish there was something at the PHP-level that could signal Kint exactly _which_ modifier it was called with, instead of parsing the source file.

On a side note, wouldn't Kint be able to differentiate between `d($a)` and `d($b)` based on the argument name?  Or in case both were called with the same argument name, differentiate based on the number of times it was called? I'm not well-versed with the internals of Kint, that's why I'm curious about it. Very interesting stuff

**Comment by jnvsor** [comment]
> On a side note, wouldn't Kint be able to differentiate between d($a) and d($b) based on the argument name?

How do you think we get the argument name? Same problem as with modifiers. In this case it'll show no name for the dumps and replace them with `$0` and `$1` in the access paths

It _can_ tell the difference between 2 calls on the same line if they have a different number of parameters though

**Edit:** Just to be clear, this is super-dirty and can only really be justified in a debug tool like this. Not knowing anything about the calling code is a pretty fundamental comp-sci principle. The reason we do it is for the param names, and the modifiers are just an extra 10% work after that

**Comment by DRSDavidSoft** [comment]
Interesting, I didn't even know even the argument names! Of course the ideal way would be for PHP to give info about the argument names, like how other platforms (such as .NET) return argument names using reflections.  Not exactly an easy way to fix this.

In case, I'm super amazed at how all this stuff works, as you said it's only justified in a debugging tool like Kint, but I must say that it does it wonderfully, and personally I've never had any issues with it.

Too sad that these kind of issues still exist, and thanks for explaining it to me. Now I understand how the argument names are extracted from the source, and why using modifiers in these scenarios wouldn't exactly make sense!

**Comment by jnvsor** [comment]
I don't work with .NET much but I'm 90% sure it doesn't offer argument names through reflection either, probably just parameter names (Which we have in PHP too)

Reflection can tell you what the parameter is named in the function definition, not what the parameter was named when it was passed into the function. In most languages that information is removed entirely when it turns into bytecode or machine code...

But now we're getting way off topic... Please try out kint-helpers and tell me if you notice anything wrong with it

**Comment by jnvsor** [comment]
@kktsvetkov @DRSDavidSoft I've published the package under version 4.0. I intend to keep the version numbers in sync for ease of use, not that there should be any real changes to kint-helpers outside of major versions

**Comment by DRSDavidSoft** [comment]
@jnvsor Good idea to keep them in sync, I haven't yet included the helpers but I'll check it out soon
